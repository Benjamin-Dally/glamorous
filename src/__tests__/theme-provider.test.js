/* eslint func-style:0 */
import React, {Component} from 'react'
import {render, mount} from 'enzyme'
import * as jestGlamorReact from 'jest-glamor-react'
import glamorous from '../'
import ThemeProvider, {CHANNEL} from '../theme-provider'

expect.extend(jestGlamorReact.matcher)
expect.addSnapshotSerializer(jestGlamorReact.serializer)

const getMockedContext = unsubscribe => ({
  [CHANNEL]: {
    getState: () => {},
    setState: () => {},
    subscribe: () => unsubscribe,
  },
})

test('renders a component with theme', () => {
  const Comp = glamorous.div(
    {
      color: 'red',
    },
    (props, theme) => ({padding: theme.padding}),
  )
  expect(
    render(
      <ThemeProvider theme={{padding: '10px'}}>
        <Comp />
      </ThemeProvider>,
    ),
  ).toMatchSnapshotWithGlamor()
})

test('theme properties updates get propagated down the tree', () => {
  class Parent extends Component {
    state = {
      padding: 10,
    }

    render() {
      return (
        <ThemeProvider theme={{padding: this.state.padding}}>
          <Child />
        </ThemeProvider>
      )
    }
  }
  const Child = glamorous.div(
    {
      color: 'red',
    },
    (props, theme) => ({padding: theme.padding}),
  )
  const wrapper = mount(<Parent />)
  expect(wrapper).toMatchSnapshotWithGlamor(`with theme prop of padding 10px`)
  wrapper.setState({padding: 20})
  expect(wrapper).toMatchSnapshotWithGlamor(`with theme prop of padding 20px`)
})

test('merges nested themes', () => {
  const One = glamorous.div({}, (props, {padding, margin}) => ({
    padding,
    margin,
  }))
  const Two = glamorous.div({}, (props, {padding, margin}) => ({
    padding,
    margin,
  }))
  expect(
    mount(
      <div>
        <ThemeProvider theme={{padding: 1, margin: 1}}>
          <div>
            <One />
            <ThemeProvider theme={{margin: 2}}>
              <Two />
            </ThemeProvider>
          </div>
        </ThemeProvider>
      </div>,
    ),
  ).toMatchSnapshotWithGlamor()
})

test('renders if children are null', () => {
  expect(
    mount(
      <ThemeProvider theme={{padding: 1}}>
        {false && <div />}
      </ThemeProvider>,
    ),
  ).toMatchSnapshotWithGlamor()
})

test('cleans up outer theme subscription when unmounts', () => {
  const unsubscribe = jest.fn()
  const context = getMockedContext(unsubscribe)
  const wrapper = mount(<ThemeProvider theme={{}} />, {context})
  wrapper.unmount()
  expect(unsubscribe).toHaveBeenCalled()
})

test('does nothing when receive same theme via props', () => {
  const theme = {margin: 2}
  const wrapper = mount(<ThemeProvider theme={theme} />)
  expect(wrapper).toMatchSnapshotWithGlamor(`with theme prop of margin 2px`)
  wrapper.setProps({theme})
  expect(wrapper).toMatchSnapshotWithGlamor(`with theme prop of margin 2px`)
})
