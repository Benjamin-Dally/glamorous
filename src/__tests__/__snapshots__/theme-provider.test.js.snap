// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merges nested themes 1`] = `
.css-mhpxxj,
[data-css-mhpxxj] {
  padding: 1px;
  margin: 1px;
}

.css-19cobm9,
[data-css-19cobm9] {
  padding: 1px;
  margin: 2px;
}

<div>
  <ThemeProvider
    theme={
      Object {
        "margin": 1,
        "padding": 1,
      }
    }
  >
    <div>
      <glamorous(div)>
        <div
          className="css-mhpxxj"
        />
      </glamorous(div)>
      <ThemeProvider
        theme={
          Object {
            "margin": 2,
          }
        }
      >
        <glamorous(div)>
          <div
            className="css-19cobm9"
          />
        </glamorous(div)>
      </ThemeProvider>
    </div>
  </ThemeProvider>
</div>
`;

exports[`renders a component with theme 1`] = `
.css-1fyph4i,
[data-css-1fyph4i] {
  color: red;
  padding: 10px;
}

<div
  class="css-1fyph4i"
/>
`;

exports[`renders if children are null 1`] = `
<ThemeProvider
  theme={
    Object {
      "padding": 1,
    }
  }
/>
`;

exports[`with theme prop of margin 2px 1`] = `
<ThemeProvider
  theme={
    Object {
      "margin": 2,
    }
  }
/>
`;

exports[`with theme prop of margin 2px 2`] = `
<ThemeProvider
  theme={
    Object {
      "margin": 2,
    }
  }
/>
`;

exports[`with theme prop of padding 10px 1`] = `
.css-1p841yy,
[data-css-1p841yy] {
  color: red;
  padding: 10px;
}

<Parent>
  <ThemeProvider
    theme={
      Object {
        "padding": 10,
      }
    }
  >
    <glamorous(div)>
      <div
        className="css-1p841yy"
      />
    </glamorous(div)>
  </ThemeProvider>
</Parent>
`;

exports[`with theme prop of padding 20px 1`] = `
.css-1e41o48,
[data-css-1e41o48] {
  color: red;
  padding: 20px;
}

<Parent>
  <ThemeProvider
    theme={
      Object {
        "padding": 20,
      }
    }
  >
    <glamorous(div)>
      <div
        className="css-1e41o48"
      />
    </glamorous(div)>
  </ThemeProvider>
</Parent>
`;
